Add Quansheng UV-K5 driver

sq5bpf opened the pull request:
UV-K5 driver, should pass tox tests

kk7ds commented:
Sorry for the delay in getting back to this. However, it looks to me like most (or all?) of my comments on the original PR haven't been addressed here? Since this got reset I just opened the two side-by-side and after the first several that look the same I stopped looking.

I spent time reviewing that in detail, so please let me know when you've got that stuff changed so I know to come back and look. Thanks!

sq5bpf commented:
For now i got rid of the C-like _getstring, removed the strict typing that you requested, removed the comment hack, and fixed the code to pass all tests (style and others).

The only thing missing is using struct.pack to construct the packets, but some other drivers are also guilty of this. I don't recall any else problems.

There are locale problems. I've regenerated them, but this will be a problem everytime something is changes in chirp main. I can regenerate it if you want.

So is using struct.pack mandatory even if other drivers are not using it, and is there anything else that would need fixing?

kk7ds commented:
Using bitfields in the memory map (instead of masks in the code) is definitely mandatory. I want you to remove all the commented out and dead code for sure. I'd really prefer you do the struct.pack stuff for the cases where you're constructing 16-bit values, yeah. For concatenating a single byte it's not as big a deal. Fixing the style issues I noted and the extraneous type conversions and things, yes please. We've got some drivers that have been hardly touched in 15 years, and surely the best practices have evolved over time. We've been working on getting some of those things addressed (like I just converted the last driver to bitwise this year, ten years after starting). I'll ask you not to create more debt we have to clean up in a brand new driver. I know it's not easy to look at a 15 year old codebase and know what the current best practices are, which is why I spent all that time going through it.

Even though it's harder to look at now, all my comments are still here for your reference:

https://github.com/kk7ds/chirp/pull/610

Thanks.
